kirkwood: switch to kernel 4.19 and use generic diag.sh

CHKDSK88 opened the pull request:
This PR includes switching kirkwood to 4.19, droping 4.14 and using generic diag.sh file.

It was compile tested on all target devices and run-tested on unofficial support of CheckPoint L-50.

CHKDSK88 committed:
kirkwood: switch to kernel 4.19

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: drop kernel 4.14 support

After switching to 4.19 kernel, 4.14 isn't needed anymore.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kirkwood: use generic diag.sh

This commit removes the target-specific diag.sh script. This way, the
generic one is used for the target, which uses DT-aliases to specify the
LEDs used.

Generic diag.sh allow to use different LEDs to indicate different states.
Non-red status LEDs for indicating boot and a running system.
Where possible, the red or orange LEDs are used to indicate failsafe
mode and a running upgrade.

Compile-tested: all target devices.
Run-tested: CheckPoint L-50

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

adrianschmutzler commented:
Hi, thanks for your PR. Your led labelling for the diag.sh changes seems a little bit inconsistent:

I typically ask for using a "led_" prefix here, so if the node is named "power" you would use "led_power: power" then, if the node is named "os-green" you would use "led_os_green: os-green" etc.
Since this scheme is not written down anywhere, though (I just tried to apply that where I review and harmonized existing code in ath79/ramips), you're not obliged to follow that.
Nevertheless, I'd consider it a good idea particularly since we are newly introducing the labels to this target, and they might be used for copy/paste later.
The way how you deal with hyphens there is good IMO. (I actually consider to also update the node names to a common scheme later.)

Despite, you seem to apply some formatting changes (trailing whitespaces and partition move/rename). I'd  prefer to have that in a separate commit, as it makes reviewing the diag.sh stuff more complicated.

CHKDSK88 commented:
@adrianschmutzler 

Thank You for the review. I drop all style changes and add `led_` prefix to led aliases.

CHKDSK88 commented:
First blocker: In EA4500 switch driver not working on 4.19.

CHKDSK88 committed:
kirkwood: fix switch dts node in EA4500 and EA3500

Changes made in switch nodes in d42c9ce commit causes problem with
correct mvsw61xx detection. This commit undo that changes.

Fixes: d42c9ce326aa ("kirkwood: add kernel 4.19 support")

Tested-by: Marcin Fedan <mfedan@gmail.com> [EA4500]
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
> First blocker: In EA4500 switch driver not working on 4.19.

Fixed.

adrianschmutzler commented:
Thanks, I consider picking the diag.sh patch for merge soon (maybe I will make it apply to 4.14 as well then).

I'm not familiar enough with the target to merge the kernel version switch myself. But since we have voted for 5.4 anyway, and having the target at 5.4 will be requirement for including it in a stable release, I don't think you will find a core-developer investing time in reviewing a 4.19 bump which is outdated by the time it is merged (even if it seems trivial). Don't get me wrong, I'm just trying to prevent you from investing even more time into this, and then seeing it's not taken up.
With the vote result as it is, I think it will be better to invest time in 5.4 support so we have time to get it stable before the next release. This is my personal perspective, though, so if you want to keep waiting for a merge it won't prevent you :-)

CHKDSK88 commented:
I made this PR to reduce number of kernel versions before 5.4. 4.19 support was ready almost one year ago.
@hauke, @bobafetthotmail  and @ynezz work at this in  @p-wassi PR #1694. I don't know why it wasn't bumped after 19.07 branch. I can't find any info.

Generic diag.sh could wait for 5.4. I know, that kirkwood isn't the most important target. ;)

adrianschmutzler commented:
> I made this PR to reduce number of kernel versions before 5.4. 4.19 support was ready almost one year ago.

I understand that, but note that we have been keeping 4.14 for updated targets anyway previously, as this is the most recent release kernel (19.07). So, I don't think bumping to 4.19 will "reduce" anything here.

Anyway, maybe we should just wait for someone else to join the discussion.

mans0n commented:
I'm planning a PR for a new kirkwood device, so I'm willing to test 4.19.

Maybe we could switch to 4.19 but keep 4.14 for now, as in other targets.

CHKDSK88 commented:
@adrianschmutzler 

I got it. First two patches will be rejected. 

Should I prepare separate PR for fix EA3500 and EA4500 for 4.19, and generic diag.sh patch for 4.14 and 4.19?

adrianschmutzler commented:
> I got it. First two patches will be rejected.

I do not think that bumping this target to 4.19 now is "wrong" for some reason. I'm just not sure whether any of the committers will be interested in reviewing/applying such a change at the moment.
I won't apply it, as my contribution to kirkwood is limited to a single patch with cosmetic changes, so I am certainly not qualified to suddenly bump kernel there.
My major aim was to stop you from investing additional time in this topic, based on my personal expectations about other committers' interest in this PR. But I may be wrong, and somebody might just merge it later.

4.14 removal might actually be obsolete based on current behavior at the moment.

The fix for EA3500 and EA4500 is completely independent from whether we bump 4.19 or not right now. I'd either leave it here or send it to the devel list separately (and still keep it here). I cannot review that one as well, so you will depend on other reviewers here, too.

Considering diag.sh you can decide for yourself what you prefer: if you want to invest the time, I'd merge a patch targeting 4.14 and 4.19 relatively quickly. But you might as well keep the current version here and wait for somebody to bump to 4.19 or redo it for 5.4 when it is ready eventually.
The only thing I won't do as it's confusing is to merge 4.19-only diag.sh now while the target is still at 4.14.

adrianschmutzler commented:
Formal remark:
Replace
`Fixes: d42c9ce`
by 12-digit hash and commit title in brackets and quotes
`Fixes: d42c9ce326aa ("kirkwood: add kernel 4.19 support")`

CHKDSK88 commented:
@adrianschmutzler 

This was my last rework of this PR.
"Remove diag.sh" patch is now first with 4.14 patches, so it can be cherry-picked now.

EA3500/4500 switch dts repair is second patch now. It could be used, or repaired in 5.4 testing version.

Better option is spend time on preparing 5.4 kernel instead 4.19.

adrianschmutzler commented:
@CHKDSK Thanks, I've picked the diag.sh into my staging tree (with some adjustments):
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

When reviewing it, I wondered what would be the best way to deal with default-state:
In your patch, it seems that you have removed default-state for the cases where it was set (locally or in kernel DTS).
However, as I understand it, it makes sense to set "default-state = on" for power/status LED, as this will light up the LED much earlier than diag.sh will do.
Despite, kirkwood uses a lot of "ucidef_set_led_default" statements in 01_leds. I'm not sure what those actually do, to me it seems like they just set the LED to "on". I assume they can be replaced by default-state in DTS with the same effect, and would just loose the user-configurability in that case.
So, to me it seems like we should put "default-state = on" in DTS for all status LEDs, and remove the corresponding statements in 01_leds, as I do not see much use in it.

What's your opinion?

If you agree, I'd prepare a commit for that myself ...

adrianschmutzler commented:
> So, to me it seems like we should put "default-state = on" in DTS for all status LEDs, and remove the corresponding statements in 01_leds, as I do not see much use in it.

I've created a patch for that and put in on top of my staging tree:

https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

mans0n commented:
@adrianschmutzler If we are adding default-state in kirkwood, are we going to apply it to other targets as well?

adrianschmutzler commented:
@mans0n So far, this is a "proposal" for kirkwood and I'm waiting for feedback on it.

When I grep for "_led_default" to see where state is still controlled by uci, I get mostly kirkwood and ar71xx. Despite: 2x x86, 6x lantiq, 1x ipq40xx.

Only kirkwood, ar71xx and x86 use it on status LEDs. Since I won't touch ar71xx for obvious reasons, the objective "do not set power/status LEDs by 01_leds, but use state-default" seems to be limited to kirkwood after all.

In contrast, using state-default for other LEDs is a different discussion, as moving config from uci to DTS will remove easy configurability of the setting, and thus will be a step backwards similar to replacing gpio-export by gpio-hogs.

Despite that, I think default-state is already used in several targets to bring the power LED up before diag.sh is working (e.g. at ath79/ramips, though it depends on reviewer ...)

bobafetthotmail commented:
> @chkdsk Thanks, I've picked the diag.sh into my staging tree (with some adjustments):
> https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging
> 
> When reviewing it, I wondered what would be the best way to deal with default-state:
> In your patch, it seems that you have removed default-state for the cases where it was set (locally or in kernel DTS).
> However, as I understand it, it makes sense to set "default-state = on" for power/status LED, as this will light up the LED much earlier than diag.sh will do.
> Despite, kirkwood uses a lot of "ucidef_set_led_default" statements in 01_leds. I'm not sure what those actually do, to me it seems like they just set the LED to "on". I assume they can be replaced by default-state in DTS with the same effect, and would just loose the user-configurability in that case.
> So, to me it seems like we should put "default-state = on" in DTS for all status LEDs, and remove the corresponding statements in 01_leds, as I do not see much use in it.
> 
> What's your opinion?

afaik most other targets do that in the dts, some (mvebu) even set secondary leds (sata/USB/whatever) in the dts. I'm ok with the idea of cleaning that up.

adrianschmutzler commented:
I'd be happy if somebody could the current state in my staging this at least on one device.

mans0n commented:
@adrianschmutzler 
AFAIK default-state turns LEDs up/down at the early stage of boot process, and that's all. We can use uci config afterward without any collision, so I don't think that's a regression.
Your patch does not only move uci config to dts, it applies to all kirkwood devices. ucidef_set_led_default on status LED is useless (it seems to be an old legacy from the past before diag.sh was added, see git log of 01_leds and diag.sh), so we can safely remove it (and we don't have to add it in dts). But if we are also touching devices which don't have uci LED config, then why don't we apply it to other targets as well? It will bring status LEDs up much earlier as you said. That's why I asked.
BTW your default-state patch works "as expected" on my yet-to-be-ported kirkwood device.

bobafetthotmail commented:
> @adrianschmutzler
> AFAIK default-state turns LEDs up/down at the early stage of boot process, and that's all. We can use uci config afterward without any collision, so I don't think that's a regression.

Afaik if you remove the line from 01_leds it will not create uci config for it, so it won't be usable from LuCi web interface until someone has added the uci config manually

adrianschmutzler commented:
> AFAIK default-state turns LEDs up/down at the early stage of boot process, and that's all. We can use uci config afterward without any collision, so I don't think that's a regression.

You're right. However, I do not see a benefit in specifying both when uci definition is kept (which is different than my argument on power LEDs), so no need to _add_ default-state for non-system LEDs.

> Your patch does not only move uci config to dts, it applies to all kirkwood devices.

Yes. If your initial comment was referring to whether just adding default-state for power LEDs on other targets makes sense, the answer is yes IMO.

> ucidef_set_led_default on status LED is useless (it seems to be an old legacy from the past before diag.sh was added, see git log of 01_leds and diag.sh),

Yes, that was my motivation.

> so we can safely remove it (and we don't have to add it in dts).

Adding it in the DTS will have the LED come up very early in boot process, way before diag.sh is triggered. That will actually helpful for the user, so he/she gets a better idea when the boot fails. I chose to add this during the third commit, I could also have done this directly in patch 1.

> But if we are also touching devices which don't have uci LED config, then why don't we apply it to other targets as well? It will bring status LEDs up much earlier as you said. That's why I asked.

Okay, now I see where you are going (maybe I thought to complicated at first :-)).

Feel free to propose patches for that, should be easy to review and merge for most (but maybe do it stepwise in case somebody complains).

adrianschmutzler commented:
> Afaik if you remove the line from 01_leds it will not create uci config for it, so it won't be usable from LuCi web interface until someone has added the uci config manually

Yes, that's why I removed those lines only for power/status LEDs, i.e. where the system controls behavior and it does not seem desirable that the user interferes (at least by default). I e.g. kept the sense LED for linksys,viper in contrast, as in this case it makes no sense to strip the user off his/her power.

mans0n commented:
> If your initial comment was referring to whether just adding default-state for power LEDs on other targets makes sense, the answer is yes IMO.

Okay, I see. I'll test it on some ramips devices.

>> ucidef_set_led_default on status LED is useless (it seems to be an old legacy from the past before diag.sh was added, see git log of 01_leds and diag.sh),
>
> Yes, that was my motivation.

Removing ucidef_set_led_default is a (cosmetic) fix, and adding default-state is another improvement (or at least a change), so I suggest splitting it into two commits.

adrianschmutzler commented:
> Removing ucidef_set_led_default is a (cosmetic) fix, and adding default-state is another improvement (or at least a change), so I suggest splitting it into two commits.

You have a point there. I've updated my staging tree accordingly. Please have another look, maybe I can merge it during the weekend ...

https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

adrianschmutzler commented:
> Okay, I see. I'll test it on some ramips devices.

Hmm, I'd have expected more:

```
adsc@buildfff:/data/openwrt$ grep -rn "default-state.*on" target/linux/ramips/dts/ | wc -l
25
adsc@buildfff:/data/openwrt$ grep -rn "default-state.*on" target/linux/ath79/dts/ | wc -l
51
```

adrianschmutzler commented:
> This was my last rework of this PR.
> "Remove diag.sh" patch is now first with 4.14 patches, so it can be cherry-picked now.
> 
> EA3500/4500 switch dts repair is second patch now. It could be used, or repaired in 5.4 testing version.
> 
> Better option is spend time on preparing 5.4 kernel instead 4.19.

@CHKDSK88 Thanks for taking care of this. I'd recommend to send patches 2 and 3 to the openwrt-devel list after I've merged the diag.sh stuff. This won't cost you too much time and there you will have a least a moderate chance that somebody just picks the patches and merges them.

I've taken the liberty to quickly rebase your patch 2 and push it to my staging tree as well. Please note that I won't merge it, but you can take it back from there and submit it directly if you want.

adrianschmutzler commented on target/linux/kirkwood/patches-4.19/104-ea3500.patch:
```diff
@@ -77,20 +77,15 @@
-+	switches {
-+		#address-cells = <1>;
-+		#size-cells = <0>;
++	mvsw61xx {
```
I suspect this should be mvsw61xx@10 as for ea4500?

CHKDSK88 replied:
My patch is reverting changes of this node, which was made in #1694.

BTW. Every dts of this node is without address (It is used only in couple devices):
````
grep -rn "mvsw61xx* {" target/linux/ 
target/linux/mvebu/patches-4.14/210-clearfog_switch_node.patch:7:+	mvsw61xx {
target/linux/mvebu/patches-4.14/003-add_switch_nodes.patch:8:+	mvsw61xx {
target/linux/mvebu/patches-4.14/003-add_switch_nodes.patch:27:+	mvsw61xx {
target/linux/mvebu/patches-4.19/210-clearfog_switch_node.patch:7:+	mvsw61xx {
target/linux/mvebu/patches-4.19/003-add_switch_nodes.patch:8:+	mvsw61xx {
target/linux/mvebu/patches-4.19/003-add_switch_nodes.patch:27:+	mvsw61xx {
target/linux/kirkwood/patches-4.14/104-ea3500.patch:80:+	mvsw61xx {
target/linux/kirkwood/patches-4.14/105-ea4500.patch:9:+	mvsw61xx {
````

adrianschmutzler replied:
> BTW. Every dts of this node is without address (It is used only in couple devices):

In principle, it should be there (DTS requires the first reg address to be appended with @reg). But I won't care about that now.

> My patch is reverting changes of this node, which was made in #1694.

Ah sorry, I didn't read carefully enough. In this case it's better to not touch the @10 at this time.

BTW: It would be interesting to provide more details on this topic (in the commit message). Unfortunately, neither your revert nor the initial commit is very chatty about that (I am not blaming you here), so at the moment it looks like a "because it works" kind of patch.

CHKDSK88 replied:
>In principle, it should be there (DTS requires the first reg address to be appended with @reg). But I won't care about that now.

I know, that is convention. IMO it should be changed in one commit for all affected dtses.

>so at the moment it looks like a "because it works" kind of patch to me.

mv61xx is platform driver. And platform device should be declared in "/" part of dts. If we add another part of tree, driver will be never probed. (Correct me if I'm wrong.)

I will add some info to commit message.

CHKDSK88 commented:
@adrianschmutzler 

Sorry for delayed answer.

>When reviewing it, I wondered what would be the best way to deal with default-state:
In your patch, it seems that you have removed default-state for the cases where it was set (locally or in kernel DTS).
However, as I understand it, it makes sense to set "default-state = on" for power/status LED, as this will light up the LED much earlier than diag.sh will do.

I thought about default-state problem. I'm not sure if removing "default-state = keep" was correct answer. Devices have status LED driven by bootloader (IMO It is most common approach). Maybe "keep" is better than "on"? If we save "keep", the LED will keep state from bootloader before diag.sh start. 

>Despite, kirkwood uses a lot of "ucidef_set_led_default" statements in 01_leds. I'm not sure what those actually do, to me it seems like they just set the LED to "on". I assume they can be replaced by default-state in DTS with the same effect, and would just loose the user-configurability in that case.

IMO we can remove status led statements from 01_leds. I think, that is functionality invented before "default state" from dts.

adrianschmutzler commented:
> Devices have status LED driven by bootloader (IMO It is most common approach). Maybe "keep" is better than "on"?

I think I actually prefer to force it "on" here, so the user can expect it to be solid-on at some point before blinking starts. With keep, you would essentially end up with undefined behavior (unless you know the deeper mechanics), wouldn't you?

adrianschmutzler commented:
@CHKDSK88 BTW: Are you fine with the use of Co-developed-by in my version of the first patch? I considered it easier than to list the individual changes... (You could e.g. use the same for your changes on the R6700v2 patch if you like).

CHKDSK88 commented:
@adrianschmutzler 

> I think I actually prefer to force it "on" here, so the user can expect it to be solid-on at some point before blinking starts. With keep, you would essentially end up with undefined behavior (unless you know the deeper mechanics), wouldn't you?

That's right. "Keep" could make led status undetermined. 

>Are you fine with the use of Co-developed-by in my version of the first patch? I considered it easier than to list the individual changes... (You could e.g. use the same for your changes on the R6700v2 patch if you like).

It is fine for me.

adrianschmutzler commented:
I've just pushed the LED patches.

mans0n commented:
@adrianschmutzler 
I just thought, if status LED turns on right after boot starts, then wouldn't it be harder to detect whether bootup goes well or wrong?
Blinking LED means that something is going on, and turned-on LED means that everything is normal. If that is the case, then perhaps it is better to turn off the LED during early stage of boot. User will know that something went wrong, if status LED does not light up even after waiting a minute.
Of course, user can know it even if we turn on status LED earlier, they just need to wait to see if the LED blinks or not. But I think it would be better if we can distinguish between the beginning and the end of boot at a glance.
FYI status LED turns off after poweroff (at least on some boards).

CHKDSK88 commented:
>Blinking LED means that something is going on, and turned-on LED means that everything is normal.

I agree with @mans0n . Solid LED light means, that everything is ok.

adrianschmutzler commented:
> I agree with @mans0n . Solid LED light means, that everything is ok.

If you want to go this way, the proper way would be to distinguish between power (default-on) and status (default-off) LEDs, as they actually imply different behavior if one's strict.

adrianschmutzler commented:
@CHKDSK88 I've taken the liberty to rebase your patches again and submit them to the list (except the 4.14 removal patch).

neheb commented:
This is superseded by https://github.com/openwrt/openwrt/pull/2799 , no?

CHKDSK88 commented:
@neheb  
> 
> This is superseded by #2799 , no?

It depends. If someone allow to use 5.4 as default kernel, this PR should be closed. If 5.4 will be used as testing kernel, 4.19 could be the main kernel and 4.14 dropped.

neheb commented:
Next release will be 5.4 based AFAIK. I don't think this should be kept around.

ynezz commented:
IMHO 4.19 seems as a derailing of the resources, I would simply focus all the efforts on 5.4 to get it in ASAP, but on the other hand if this is ready, then fine with me 

BTW that `kirkwood: switch to kernel 4.19` commit is missing description, other than that 'Acked-by: Petr Štetiar <ynezz@true.cz>'

ynezz commented:
2nd BTW, I wouldn't remove that 4.14 kernel right now, I would leave it there and remove it together with other 4.14 kernel cleanups

adrianschmutzler commented:
Since the 5.4 PR has been accepted, I close this now.
If target is moved to 5.4 directly, one could even consider removing the 4.19 support in this target entirely.
